frontend/src/tests/import-export.test.js

```diff
@@ -190,4 +185,622 @@ describe("Import/export library functionality", () => {
+                applicants: [
+                    // applicant to be modified
+                    {
+                        first_name: "John",
+                        last_name: "Doe",
+                        utorid: "johnd",
+                        email: "goofy-duck@donald.com",
+                        student_number: "OLD10000000",
+                    },
+                    // applicant to be duplicated
+                    {
+                        first_name: "哈利",
+                        last_name: "Potter",
+                        utorid: "potterh",
+                        email: "harry@potter.com",
+                        student_number: "999666999",
+                        phone: "41888888888",
+                    },
+                ],
+            }
+        );
+        expect(applicantsDiff).toMatchSnapshot();
+    });
+});
+
+it("Import Positions from JSON/CSV/XLSX", () => {
+    // import correct positions from XLSX
+    let normalizedSpreadsheetPositions = normalizeImport(
+        {
+            fileType: "spreadsheet",
+            data: parseSpreadsheet("positions_correct.xlsx"),
+        },
+        positionSchema
+    );
+    expect(normalizedSpreadsheetPositions).toMatchSnapshot();
+    // import correct positions from JSON
+    const normalizedJsonPositions = normalizeImport(
+        {
+            fileType: "json",
+            data: objectJSON["positions"],
+        },
+        positionSchema
+    );
+    expect(normalizedJsonPositions).toMatchSnapshot();
+    // import positions data missing required key utorid should throw error
+    expect(() =>
+        normalizeImport(
+            {
+                fileType: "spreadsheet",
+                data: parseSpreadsheet("positions_missing_required_keys.xlsx"),
+            },
+            positionSchema
+        ).toThrow(Error)
+    );
+    // import positions data with invalid start_date should throw error
+    expect(() =>
+        normalizeImport(
+            {
+                fileType: "spreadsheet",
+                data: parseSpreadsheet("positions_invalid_date_columns.xlsx"),
+            },
+            positionSchema
+        ).toThrow(Error)
+    );
+    for (const item of normalizedSpreadsheetPositions) {
+        item.instructors = diffImport
+            .instructorsListFromField(item.instructors || [], {
+                instructors: [
+                    {
+                        first_name: "Henry",
+                        last_name: "Smith",
+                        utorid: "smithh",
+                    },
+                    {
+                        first_name: "Emily",
+                        last_name: "Garcia",
+                        utorid: "garciae",
+                    },
+                ],
+            })
+            .map((x) => x.utorid);
+    }
+    // compute the difference with existing positions
+    const positionsDiff = diffImport.positions(normalizedSpreadsheetPositions, {
+        positions: [
```

Comment:
You have some data stored in external `json` files and some of it inline. Why?

Reply:
all the inline `json`s are used as "context" in `diffImport` functions, which is considered as the "existing" objects and is compared with the imported objects to find the difference. in this case, should I also store these "context" objects into files or a explanation is enough (keep them inline)?

Reply:
Do you think it helps readability to have this JSON inline?

Reply:
that makes sense... moved those to a collection of import object contexts